extensions/amp-tiktok/0.1/amp-tiktok.js

```diff
@@ -0,0 +1,167 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout, isLayoutSizeDefined} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {
+  childElementByTag,
+  createElementWithAttributes,
+  removeElement,
+} from '../../../src/dom';
+import {debounce} from '../../../src/utils/rate-limit';
+import {getData, listen} from '../../../src/event-helper';
+import {resetStyles, setStyle, setStyles} from '../../../src/style';
+import {tryParseJson} from '../../../src/json';
+
+export class AmpTiktok extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {!Element} */
+    this.iframe_ = null;
+
+    /** @private {string} */
+    this.videoId_ = null;
+
+    /** @private {?Function}*/
+    this.unlistenMessage_ = null;
+
+    this.resizeOuter_ = debounce(
+      this.win,
+      (height) => {
+        resetStyles(this.iframe_, [
+          'width',
+          'height',
+          'position',
+          'opacity',
+          'pointer-events',
+        ]);
+        this.iframe_.removeAttribute('aria-hidden');
+        this.iframe_.setAttribute('aria-title', 'Tiktok');
+        this.applyFillContent(this.iframe_);
+        this.forceChangeHeight(height);
+      },
+      1000
+    );
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    //See
+    //https://developers.tiktok.com/doc/Embed
+    Services.preconnectFor(this.win).url(
+      this.getAmpDoc(),
+      'https://www.tiktok.com',
+      opt_onLayout
+    );
+  }
+
+  /** @override */
+  buildCallback() {
+    const {src} = this.element.dataset;
+    if (src) {
+      this.videoId_ = src.replace(/^((.+\/)?)(\d+)\/?$/, (_, _1, _2, id) => id);
+    } else {
+      const blockquoteOrNull = childElementByTag(this.element, 'blockquote');
+      if (
+        !blockquoteOrNull ||
+        !blockquoteOrNull.hasAttribute('placeholder') ||
+        !blockquoteOrNull.dataset.videoId
+      ) {
+        return;
+      }
+      this.videoId_ = blockquoteOrNull.dataset.videoId;
+    }
+  }
+
+  /** @override */
+  layoutCallback() {
+    const iframe = this.element.ownerDocument.createElement('iframe');
+    const {locale = 'en-US'} = this.element.dataset;
+    const src = `https://www.tiktok.com/embed/v2/${encodeURIComponent(
+      this.videoId_
+    )}?lang=${encodeURIComponent(locale)}`;
+    this.iframe_ = iframe;
+
+    this.unlistenMessage_ = listen(
+      this.win,
+      'message',
+      this.handleTiktokMessages_.bind(this)
+    );
+
+    this.iframe_.src = src;
+    this.iframe_.setAttribute('name', '__tt_embed__v$');
+    this.iframe_.setAttribute('aria-hidden', 'true');
+    this.iframe_.setAttribute('frameborder', '0');
+    this.iframe_.setAttribute('class', 'i-amphtml-tiktok-centered');
```

Comment:
1. Use `createElementWithAttributes`.
2. We don't want to center the video just yet, that happens on final resize instead of the `applyFillContent` above.
3. This classname is not doing anything because there's a typo.

```diff
- in css: i-amphtml-tikok-centered
+ in js:  i-amphtml-tiktok-centered
```